File: server/pkg/model/conversation.go
package model

type ConvType int

const (
	ConvP2P    ConvType = 1
	ConvGroup  ConvType = 2
	ConvSystem ConvType = 3
)

type ConvRole int

const (
	ConvRoleMember ConvRole = 0
	ConvRoleAdmin  ConvRole = 1
	ConvRoleOwner  ConvRole = 2
)

type Conversation struct {
	ConvID     string         `json:"conv_id"`
	Type       ConvType       `json:"type"`
	Name       string         `json:"name"`
	OwnerID    string         `json:"owner_id"`
	Avatar     string         `json:"avatar,omitempty"`
	Cover      string         `json:"cover,omitempty"`
	Notice     string         `json:"notice,omitempty"`
	Headline   string         `json:"headline,omitempty"`
	MaxMembers int            `json:"max_members,omitempty"`
	LastMsgID  int64          `json:"last_msg_id,omitempty"`
	LastMsgAt  int64          `json:"last_msg_at,omitempty"`
	CreatedAt  int64          `json:"created_at"`
	Settings   map[string]any `json:"settings,omitempty"`
}

type ConvMember struct {
	ConvID   string   `json:"conv_id"`
	UserID   string   `json:"user_id"`
	Role     ConvRole `json:"role"`
	Nickname string   `json:"nickname,omitempty"`
	Mute     bool     `json:"mute"`
	JoinedAt int64    `json:"joined_at"`
	UserType UserType `json:"user_type"`
	WakeMode WakeMode `json:"wake_mode"`
	Pinned   bool     `json:"pinned"`
}

type JoinRequestStatus int

const (
	JoinRequestPending  JoinRequestStatus = 0
	JoinRequestApproved JoinRequestStatus = 1
	JoinRequestRejected JoinRequestStatus = 2
)

type JoinRequest struct {
	ConvID    string            `json:"conv_id"`
	UserID    string            `json:"user_id"`
	Status    JoinRequestStatus `json:"status"`
	CreatedAt int64             `json:"created_at"`
	UpdatedAt int64             `json:"updated_at"`
}

File: server/web/src/types/conversation.ts
import type { UserType, WakeMode } from './user'

export enum ConvType {
  P2P = 1,
  Group = 2,
  System = 3,
}

export enum ConvRole {
  Member = 0,
  Admin = 1,
  Owner = 2,
}

export enum JoinRequestStatus {
  Pending = 0,
  Approved = 1,
  Rejected = 2,
}

export interface Conversation {
  conv_id: string
  type: ConvType
  name: string
  owner_id: string
  avatar?: string
  cover?: string
  primary_color?: string
  max_members?: number
  last_msg_id?: number
  last_msg_at?: number
  created_at: number
}

export interface ConvMember {
  conv_id: string
  user_id: string
  role: ConvRole
  nickname?: string
  mute: boolean
  joined_at: number
  user_type: UserType
  wake_mode: WakeMode
}

export interface JoinRequest {
  conv_id: string
  user_id: string
  status: JoinRequestStatus
  created_at: number
  updated_at: number
}

// Client-side conversation list item (matches server's ConvListItem)
export interface LastMessageInfo {
  msg_id: number
  sender_id: string
  sender_name: string
  body: string
  content_type: number
  timestamp: number
  status: number
}

export interface ConvListItem {
  conv_id: string
  type: ConvType
  name: string
  avatar: string
  unread_count: number
  last_message?: LastMessageInfo
  last_msg_at: number
  role: number
  mute: boolean
  mention_me: boolean
  partner_type: number
  pinned: boolean
}

export interface ConversationDetail extends Conversation {
  members: ConvMember[]
  notice?: string
  share_token?: string
}
